### PR TITLE
CIT-459: Deterministic batching algorithm

### PR DESCRIPTION
**Pull request by ChrisGreenaway**
## Proposed changes

Implement the batching algorithm described here: https://github.com/albinsuresh/thin-edge.io-specs/blob/main/src/telemetry-data/message-batching/message-batching.md

## Types of changes

What types of changes does your code introduce to `thin-edge.io`?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactorings that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
https://cumulocity.atlassian.net/browse/CIT-459
<br/>

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [ ] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [ ] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: common/batcher/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+[package]
+name = "batcher"
+version = "0.1.0"
+authors = ["Chris Greenaway <chris.greenaway@softwareag.com>"]
```
Comment (PradeepKiruvale) [inline common/batcher/Cargo.toml:4]: May be need to change the authors to "Software AG <thin-edge-team@softwareag.com>

**Review by PradeepKiruvale — COMMENTED** [review]
Left some comments.

File: common/batcher/Cargo.toml (hunk 1 above)
Comment (ChrisGreenaway) [reply common/batcher/Cargo.toml:4]: I will remove it completely.

File: common/batcher/src/batch.rs [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+        let new_batch_end = self.batch_end;
+
+        let all_events = std::mem::take(&mut self.events);
+        self.batch_end = split_point;
+
+        for event in all_events
+            .into_iter()
+            .map(|(_key, event)| event)
+            .chain(once(event))
+        {
```
[the remark this answers is not in the document]
Comment (ChrisGreenaway) [reply common/batcher/src/batch.rs:70]: We go over all the events in this batch plus the new event and allocate them into either the existing batch or the new batch.

File: common/batcher/src/batcher.rs [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+            .max_by(|batch1, batch2| batch1.batch_end().cmp(&batch2.batch_end()))
+    }
+
+    fn next_batch(&self, event_time: DateTime<Utc>) -> Option<&Batch<B>> {
+        self.batches
+            .iter()
+            .filter(|batch| batch.batch_start() > event_time)
+            .min_by(|batch1, batch2| batch1.batch_start().cmp(&batch2.batch_start()))
+    }
+}
```
[the remark this answers is not in the document]
Comment (ChrisGreenaway) [reply common/batcher/src/batcher.rs:140]: They were in the lib.rs, but I will move them into batcher.rs.

I don't intend to test `previous_batch` and `next_batch` methods directly as they are internal implementation details.

File: common/batcher/src/driver.rs [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+
+        let event1 = TestBatchEvent::new(1, Utc::now());
+        input_send.send(BatchDriverInput::Event(event1)).await?;
+        sleep(Duration::from_millis(100)).await;
+        let event2 = TestBatchEvent::new(2, Utc::now());
+        input_send.send(BatchDriverInput::Event(event2)).await?;
+
+        assert_recv_batch(&mut output_recv, vec![event1]).await;
+        assert_recv_batch(&mut output_recv, vec![event2]).await;
+
```
[the remark this answers is not in the document]
Comment (ChrisGreenaway) [reply common/batcher/src/driver.rs:169]: The timer on the first batch expires, so the first batch is sent on the output channel, then the second event arrives and goes in a second batch. The second batch expires, so it too gets sent on the output channel.

However, I think the test as written was racy and inefficient, so I'll get rid of the sleep and replace it with a wait for the first batch.

File: common/batcher/src/driver.rs [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+        Ok(())
+    }
+
+    async fn assert_recv_batch(
+        output_recv: &mut Receiver<BatchDriverOutput<TestBatchEvent>>,
+        expected: Vec<TestBatchEvent>,
+    ) {
+        match timeout(Duration::from_secs(10), output_recv.recv()).await {
+            Ok(Some(BatchDriverOutput::Batch(batch))) => assert_batch(batch, expected),
+            other => panic!("Failed to receive batch: {:?}", other),
```
[the remark this answers is not in the document]
Comment (ChrisGreenaway) [reply common/batcher/src/driver.rs:179]: other isn't a special keyword, it's just a variable name. That branch matches anything, so `other` will contain anything other than `Ok(Some(BatchDriverOutput::Batch()))`.

**Review by makr11st — APPROVED** [review]

**Pull request merged**